cpp/src/groupby/new_groupby.cu

```diff
@@ -219,3 +226,140 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
+  gdf_size_type nrows = input_table.num_rows();
+
+  rmm::device_vector<void*> d_cols(ncols);
+  rmm::device_vector<int> d_types(ncols, 0);
+  void** d_col_data = d_cols.data().get();
+  int* d_col_types = d_types.data().get();
+
+  bool nulls_are_smallest = (context->flag_null_sort_behavior == GDF_NULL_AS_SMALLEST);
+
+  gdf_index_type* result_end;
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+  auto exec = rmm::exec_policy(stream)->on(stream);
+
+  gdf_index_type* unique_indices;
+  RMM_TRY( RMM_ALLOC(&unique_indices, sizeof(gdf_index_type) * nrows, stream) );
+
+  bool const have_nulls{ std::any_of(input_table.begin(), input_table.end(), [](gdf_column const* col){ return col->null_count > 0;}) };
+  if (have_nulls){
+    rmm::device_vector<gdf_valid_type*> d_valids(ncols);
+    gdf_valid_type** d_valids_data = d_valids.data().get();
+
+    soa_col_info(input_table.begin(), ncols, d_col_data, d_valids_data, d_col_types);
+
+    LesserRTTI<gdf_size_type> comp(d_col_data, d_valids_data, d_col_types, nullptr, ncols, nulls_are_smallest);
+
+    auto counting_iter = thrust::make_counting_iterator<gdf_size_type>(0);
+
+    result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows,
+                              unique_indices,
+                              [comp]  __device__(gdf_size_type key1, gdf_size_type key2){
+                              return comp.equal_with_nulls(key1, key2);
+                            });
+
+  } else {
+    soa_col_info(input_table.begin(), ncols, d_col_data, nullptr, d_col_types);
+
+    LesserRTTI<gdf_size_type> comp(d_col_data, nullptr, d_col_types, nullptr, ncols, nulls_are_smallest);
+
+    auto counting_iter = thrust::make_counting_iterator<gdf_size_type>(0);
+
+    result_end = thrust::unique_copy(exec, counting_iter, counting_iter+nrows,
+                              unique_indices,
+                              [comp]  __device__(gdf_size_type key1, gdf_size_type key2){
+                              return comp.equal(key1, key2);
+                            });
+  }
+
+  gdf_size_type new_sz = thrust::distance(unique_indices, result_end);
+
+  cudaStreamSynchronize(stream);
+  cudaStreamDestroy(stream);
+
+  return std::make_tuple(unique_indices, new_sz);
+}
+
+std::tuple<cudf::table, gdf_index_type*, gdf_size_type> 
+gdf_group_by_without_aggregations(cudf::table const& input_table,
+                                  gdf_size_type num_key_cols,
+                                  gdf_index_type const * key_col_indices,
+                                  gdf_context* context)
+{
+  CUDF_EXPECTS(nullptr != key_col_indices, "key_col_indices is null");
+  CUDF_EXPECTS(0 < num_key_cols, "number of key colums should be greater than zero");
+
+  if (0 == input_table.num_rows()) {
+    return std::make_tuple(cudf::table(), nullptr, 0);
+  }
+
+  gdf_size_type nrows = input_table.num_rows();
+  
+  // Allocate output columns
+  std::vector<gdf_dtype> output_types(input_table.num_columns());
```

Comment:
I had a helper function to make this operation easier in another PR. I separated it out so you can use is here.

See https://github.com/rapidsai/cudf/pull/1617/files#diff-c00c04072fc8516510376c2d9a71bb7cR138

Reply:
done